src/actor/mod.rs

```diff
@@ -0,0 +1,98 @@
+mod module;
+mod wasmbus;
+
+pub use module::{Instance as ModuleInstance, Module};
+
+use core::fmt::{self, Debug};
+use core::ptr::NonNull;
+
+use std::sync::Arc;
+
+use anyhow::{ensure, Context, Result};
+use wascap::jwt;
+use wascap::wasm::extract_claims;
+
+mod wasm {
+    #[allow(non_camel_case_types)]
+    pub type ptr = u32;
+    #[allow(non_camel_case_types)]
+    pub type usize = u32;
+
+    pub const ERROR: usize = usize::MAX;
+    pub const SUCCESS: usize = 1;
+}
+
+mod guest_call {
+    use super::{wasm, NonNull};
+
+    pub type Params = (wasm::usize, wasm::usize);
+    pub type Result = wasm::usize;
+
+    pub type State = (NonNull<[u8]>, NonNull<[u8]>);
+}
+
+/// Extracts and validates claims contained within `WebAssembly` module
+fn actor_claims(wasm: impl AsRef<[u8]>) -> Result<jwt::Claims<jwt::Actor>> {
+    let claims = extract_claims(wasm)
+        .context("failed to extract module claims")?
+        .context("execution of unsigned Wasm modules is not allowed")?;
+    let v = jwt::validate_token::<jwt::Actor>(&claims.jwt)
+        .context("failed to validate module token")?;
+    ensure!(!v.expired, "token expired at `{}`", v.expires_human);
+    ensure!(
+        !v.cannot_use_yet,
+        "token cannot be used before `{}`",
+        v.not_before_human
+    );
+    ensure!(v.signature_valid, "signature is not valid");
+
+    Ok(claims.claims)
+}
+
+struct Ctx<'a, H> {
+    wasi: wasmtime_wasi::WasiCtx,
+    wasmbus: wasmbus::Ctx<H>,
+    claims: &'a jwt::Claims<jwt::Actor>,
+}
+
+impl<H> Debug for Ctx<'_, H> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Ctx")
+            .field("runtime", &"wasmtime")
+            .field("wasmbus", &self.wasmbus)
+            .field("claims", &self.claims)
+            .finish()
+    }
+}
+
+impl<'a, H> Ctx<'a, H> {
+    fn new(claims: &'a jwt::Claims<jwt::Actor>, handler: Arc<H>) -> Result<Self> {
```

Comment:
per discussion, we wanted to remove per-actor handlers.

Reply:
Those are removed, this is an internal constructor